### PR TITLE
Replace assertion errors on IR assumption violation

### LINKED ISSUE #7217: Regression: SSA scope check assertion error
If anyone see `AssertionError` from:

https://github.com/numba/numba/blob/master/numba/core/ssa.py#L270

please report to this issue with reproducers.

### PR DESCRIPTION
**Pull request by sklam**
with pedantic warnings.

Closes #7217

### CODE REVIEW

File: numba/core/ssa.py [hunk 1]
```diff
@@ -267,7 +267,12 @@ def on_assign(self, states, assign):
             if len(defmap) == 0:
                 newtarget = assign.target
                 _logger.debug("first assign: %s", newtarget)
-                assert newtarget.name in scope.localvars
+                if newtarget.name not in scope.localvars:
+                    wmsg = (
+                        f"variable name {newtarget.name!r} not in scope "
+                        f"in {assign} on line {assign.loc}",
+                    )
+                    warnings.warn(errors.NumbaIRAssumptionWarning(wmsg))
```
Comment (stuartarchibald) [inline numba/core/ssa.py:275]:
```suggestion
                if newtarget.name not in scope.localvars:
                    wmsg = f"variable name {newtarget.name!r} not in scope"
                    warnings.warn(errors.NumbaIRAssumptionWarning(wmsg, loc=assign.loc))
```
Does it help to use Numba's error message `loc` handling here?

Comment (sklam) [reply numba/core/ssa.py:275]: oh, i forgot that it has that

Comment (sklam) [reply numba/core/ssa.py]: changed in https://github.com/numba/numba/pull/7223/commits/ec01acdbf837795b410bfce0cb238b5b3233a207

**Comment by stuartarchibald** [comment]
Note: verified behaviour in ec01acd against https://github.com/numba/numba/issues/7217#issuecomment-884293457

File: numba/core/ssa.py [hunk 2]
```diff
@@ -267,7 +267,10 @@ def on_assign(self, states, assign):
             if len(defmap) == 0:
                 newtarget = assign.target
                 _logger.debug("first assign: %s", newtarget)
-                assert newtarget.name in scope.localvars
+                if newtarget.name not in scope.localvars:
+                    wmsg = f"variable name {newtarget.name!r} not in scope."
```
Comment (stuartarchibald) [inline numba/core/ssa.py:271]:
```suggestion
                    wmsg = f"variable {newtarget.name!r} is not in scope."
```

File: numba/tests/test_ir.py [hunk 3]
```diff
@@ -466,5 +484,76 @@ def baz(a, b):
         check_diffstr(tmp, ["c + b", "b + c"])
 
 
+class TestIRPedanticChecks(TestCase):
+    def test_var_in_scope_assumption(self):
+        # Create a pass that clear ir.Scope in ir.Block
```
Comment (stuartarchibald) [inline numba/tests/test_ir.py:489]:
```suggestion
        # Create a pass that clears ir.Scope in ir.Block
```

File: numba/tests/test_ir.py [hunk 4]
```diff
@@ -466,5 +484,76 @@ def baz(a, b):
...
+                FunctionPass.__init__(self)
+
+            # implement method to do the work, "state" is the internal compiler
+            # state from the CompilerBase instance.
+            def run_pass(self, state):
+                func_ir = state.func_ir
+                # walk the blocks
+                for blk in func_ir.blocks.values():
+                    oldscope = blk.scope
+                    # put in a empty Scope
```
Comment (stuartarchibald) [inline numba/tests/test_ir.py:504]:
```suggestion
                    # put in an empty Scope
```

File: numba/tests/test_ir.py [hunk 5]
```diff
@@ -466,5 +484,76 @@ def baz(a, b):
...
+                    blk.scope = ir.Scope(parent=oldscope.parent,
+                                         loc=oldscope.loc)
+                return True
+
+        # Create a pass that always fail to stop the compiler
```
Comment (stuartarchibald) [inline numba/tests/test_ir.py:509]:
```suggestion
        # Create a pass that always fails, to stop the compiler
```

File: numba/tests/test_ir.py [hunk 6]
```diff
@@ -466,5 +484,76 @@ def baz(a, b):
...
+        @register_pass(mutates_CFG=False, analysis_only=False)
+        class FailPass(FunctionPass):
+            _name = "_fail"
+
+            def __init__(self, *args, **kwargs):
+                FunctionPass.__init__(self)
+
+            def run_pass(self, state):
+                # This is unreachable. SSA pass should have raised before this
+                # pass
```
Comment (stuartarchibald) [inline numba/tests/test_ir.py:519]:
```suggestion
                # This is unreachable. SSA pass should have raised before this
                # pass when run with `error.NumbaPedanicWarning`s raised as
                # errors.
```

File: numba/tests/test_ir.py [hunk 7]
```diff
@@ -466,5 +484,76 @@ def baz(a, b):
...
+                pm.add_pass(TranslateByteCode, "analyzing bytecode")
+                pm.add_pass(IRProcessing, "processing IR")
+                pm.add_pass(RemoveVarInScope, "_remove_var_in_scope")
+                pm.add_pass(ReconstructSSA, "ssa")
+                pm.add_pass(FailPass, "_fail")
+                pm.finalize()
+                return [pm]
+
+        # import logging
+        # logging.basicConfig(level=logging.DEBUG)
```
Comment (stuartarchibald) [inline numba/tests/test_ir.py:534]: Dead comments

File: numba/core/errors.py [hunk 8]
```diff
@@ -82,6 +82,18 @@ class NumbaInvalidConfigWarning(NumbaWarning):
...
+
+class NumbaPedanticWarning(NumbaWarning):
+    """
+    Warning category for reporting pedantic messages.
+    """
+
+
+class NumbaIRAssumptionWarning(NumbaPedanticWarning):
+    """
+    Warning category for reporting IR assumption violation.
```
Comment (stuartarchibald) [inline numba/core/errors.py:94]:
```suggestion
    Warning category for reporting an IR assumption violation.
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. There's a few minor things to resolve but looks good else.

File: numba/tests/test_ir.py [hunk 9]
```diff
@@ -466,5 +484,77 @@ def baz(a, b):
...
```
Comment (stuartarchibald) [inline numba/tests/test_ir.py:535]:
```suggestion
```

File: numba/tests/test_ir.py [hunk 10]
```diff
@@ -466,5 +484,77 @@ def baz(a, b):
...
+        with warnings.catch_warnings():
+            # Make NumbaPedanticWarning an error
+            warnings.simplefilter("error", errors.NumbaPedanticWarning)
+            # Catch NumbaIRAssumptionWarning
+            with self.assertRaises(errors.NumbaIRAssumptionWarning) as raises:
+                dummy(1)
+            # Verify the error message
+            self.assertRegex(
+                str(raises.exception),
+                r"variable name '[a-z]' not in scope",
```
Comment (stuartarchibald) [inline numba/tests/test_ir.py:555]:
```suggestion
                r"variable '[a-z]' is not in scope",
```

**Review by stuartarchibald — COMMENTED** [review]
Couple of suggestions to hopefully fix the build.

**Comment by sklam** [comment]
Btw, i'm trying out if the warning message should suggest users to report.

Currently, the warning looks like:

```python
/path/to/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable 'a' is not in scope.
  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
/path/to/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable 'b' is not in scope.
  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
```

With the suggestion:

```python
/path/to/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable 'b' is not in scope.
This error came from an internal pedantic check. Please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new

  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
/path/to/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable 'a' is not in scope.
This error came from an internal pedantic check. Please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new

  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
```

The long message can be very annoying.

**Comment by stuartarchibald** [comment]
RE https://github.com/numba/numba/pull/7223#issuecomment-884504472 Whilst the long messages are not great, I'd hope that the frequency at which this is hit is really small else it indicates a lot more problems. I guess this comes down to our confidence around having got Numba's internals sufficiently fixed up for this to be a rarity and definitely something that should be reported vs. an annoyance?

File: numba/tests/test_ir.py [hunk 11]
```diff
@@ -466,5 +484,75 @@ def baz(a, b):
...
+        @register_pass(mutates_CFG=False, analysis_only=False)
+        class FailPass(FunctionPass):
+            _name = "_fail"
+
+            def __init__(self, *args, **kwargs):
+                FunctionPass.__init__(self)
+
+            def run_pass(self, state):
+                # This is unreachable. SSA pass should have raised before this
+                # pass when run with `error.NumbaPedanicWarning`s raised as
```
Comment (stuartarchibald) [inline numba/tests/test_ir.py:519]:
```suggestion
                # pass when run with `error.NumbaPedanticWarning`s raised as
```

Comment (sklam) [reply numba/tests/test_ir.py]: fixed

File: numba/core/errors.py [hunk 12]
```diff
@@ -326,6 +328,13 @@ def termcolor():
             _termcolor_inst = HighlightColorScheme(scheme)
         return _termcolor_inst
 
+
+pedantic_warning_info = """
+This error came from an internal pedantic check. Please report the error message
```
Comment (stuartarchibald) [inline numba/core/errors.py:333]:
```suggestion
This warning came from an internal pedantic check. Please report the warning message
```

Comment (sklam) [reply numba/core/errors.py]: fixed

**Comment by sklam** [comment]
@esc, can you review this? It's just need a last look.

File: numba/tests/test_ir.py [hunk 13]
```diff
@@ -466,5 +484,75 @@ def baz(a, b):
...
+                pm.add_pass(TranslateByteCode, "analyzing bytecode")
+                pm.add_pass(IRProcessing, "processing IR")
+                pm.add_pass(RemoveVarInScope, "_remove_var_in_scope")
+                pm.add_pass(ReconstructSSA, "ssa")
+                pm.add_pass(FailPass, "_fail")
+                pm.finalize()
+                return [pm]
+
+        @njit(pipeline_class=MyCompiler)
+        def dummy(x):
```
Comment (esc) [inline numba/tests/test_ir.py:535]: Perhaps add a comment explaining why this `dummy` is needed: multiple assignments to  the same variable name are needed to trigger SSA.

**Review by esc — CHANGES_REQUESTED** [review]
A few minor things to edit, ready to merge after that.

File: numba/tests/test_ir.py (hunk 13 above)
Comment (sklam) [reply numba/tests/test_ir.py:535]: done in https://github.com/numba/numba/pull/7223/commits/3d6fb07627fa081004d73f1378dfb569a90699ad

File: numba/core/errors.py [hunk 14]
```diff
@@ -314,6 +328,13 @@ def termcolor():
...
+
+pedantic_warning_info = """
+This warning came from an internal pedantic check. Please report the warning
+message and traceback, along with a minimal reproducer at:
+https://github.com/numba/numba/issues/new?template=bug_report.md
+"""
+
 feedback_details = """
 Please report the error message and traceback, along with a minimal reproducer
 at: https://github.com/numba/numba/issues/new
```
Comment (esc) [inline numba/core/errors.py:340]: Do you want to update this URL too, while there?

**Review by esc — APPROVED** [review]
Changes are fine, I left one comment but that is perhaps something for a future PR.

File: numba/core/errors.py (hunk 14 above)
Comment (sklam) [reply numba/core/errors.py:340]: Let's do this in a future PR.

Comment (sklam) [reply numba/core/errors.py:340]: i've made a task issue for it: https://github.com/numba/numba/issues/7261

**Pull request merged**